### DISCUSSION TITLE
How to calibrate offsets and gain?

### DISCUSSION THREAD

**Discussion by carstensun** [OpenHantek/OpenHantek6022#276] [Q&A] [answered]
Please excuse my complete newbie question. This my first time on GitHub and first time asking a question. I am guessing however that I am not the only one with this problem.
I am a happy user of OpenHantek6022 on Linux Mint 20.2 and am very grateful for the tremendous work put into it and so generously shared.

I see offsets on the channels and after spending days researching how to calibrate I am at my wits end with how to do it.

I have found https://github.com/Ho-Ro/Hantek6022API and calibrate_6022.py and https://github.com/Ho-Ro/Hantek6022API/blob/main/README.md#create-calibration-values-for-openhantek
There are beautiful thorough descriptions of the process, however I cannot see anywhere how I get to the point of being able to run calibrate_6022.py.
Under "Developed under Linux" the first instruction is "sudo cp 60-hantek-6022-usb.rules /etc/udev/rules.d/", but how do I get "60-hantek-6022-usb.rules" and put it in the right place?
I have put my question here under OpenHantek6022, because this is where I am, having the problem, as a user of OpenHantek6022. I am guessing I need to somehow download and install something from Hantek6022API before I can run calibrate_6022.py
It looks like I can find instructions on how to do it on Windows 10, but I cannot see anywhere how to do it on Linux. I have read the OpenHantek6022 manual without finding the answer. If another manual or something else answers my question, please point me there so I can RTFM  ;o)

So please help me with a scope user's step by step guide to calibrate the 6022. I have a 6022BL and am running Openhantek6022 on Linux Mint 20.2 Cinnamon.

**Comment by Ho-Ro** [comment, marked as answer]
> ... I cannot see anywhere how I get to the point of being able to run calibrate_6022.py.
### TL;DR:
install the lates [`*.deb` package](https://github.com/Ho-Ro/Hantek6022API/releases/download/2.10.1/hantek6022api_2.10.1-1_all.deb) and type `calibrate_6022.py` in a terminal window.

### Longer explanation:
Hantek6022API started as a low-level project for developers, mainly targeted on the development of firmware (e.g. for OpenHantek6022) and during the years I wrote some tools also interesting for the user of the Hantek scope HW.

> Under "Developed under Linux" the first instruction is "sudo cp 60-hantek-6022-usb.rules /etc/udev/rules.d/", but how do I get "60-hantek-6022-usb.rules" and put it in the right place?

The typical GitHub **development process** starts by cloning from GitHub to your local computer. Get the link to the GitHub repo from the green `Code` button:
![image](https://user-images.githubusercontent.com/12542359/156995938-8cd6e3cb-e1a0-4bb9-a0e5-30519aefa647.png)
and type `git clone https://github.com/Ho-Ro/Hantek6022API.git`. Then enter the directory Hantek6022API and work inside this directory, there it is under `udev`.
If you do not want to develop and didn't clone the project and you are just looking at the [GitHub project page](https://github.com/Ho-Ro/Hantek6022API) (this is where the development and discussion takes place), please click [udev](https://github.com/Ho-Ro/Hantek6022API/tree/main/udev), there you will find the udev file, but this step redundant, because you already got the right permissions by installing *OpenHantek6022* (you did it via the `*.deb` package, didn't you?). 

If you read further down from [Developed under Linux](https://github.com/Ho-Ro/Hantek6022API#developed-under-linux) you will come to the line saying: 
> To build the custom firmware run `make` in the top level directory

Start reading there, until it mentions: 

>  This installs the python modules together with some utility programs. 

This was the complete build and install process for **developer** - as a Mint user you do not need to build your own packages, just install the debian package.

The step by step installation tutorial for Hantek6022API:

Download the latest `*.deb` package from **Releases** on the right
![image](https://user-images.githubusercontent.com/12542359/156991478-9ec5ab94-c2de-4e0b-9b54-0d58a0ff8914.png)
and install it via your package management (open a terminal window, go to the download directory and type `sudo apt install ./hantek6022api*.deb`) - you're done. It installs the python package as well as some utilities from directory [examples](https://github.com/Ho-Ro/Hantek6022API/tree/main/examples), (all tools with a name of `*_6022.py`). You can run them by typing the name in your preferred terminal. 
If you want to play with the python tools, just copy one of the examples into an empty directory and run it from there (either by typing `python name_of_the_tool.py` or by just calling it by `./name_of_the_tool.py`, make changes, etc. No need to type `make` or install something again.
Compared to Windows the Linux installation is as easy as 1-2-3, therefor you find a lot of Win-related discussion and hints and much less for Linux.
But anyhow, thank you for reporting your issues, I will update the [README](https://github.com/Ho-Ro/Hantek6022API/blob/main/README.md) accordingly.

**Reply by carstensun** [reply]
Wow, thank you for an excellent answer. Thank you for taking the time and care to explain it so well. Your one line instruction, with link to the current package to install, did it for me in about 10-15 min from reading it until the the scope had been offset calibrated. It will be another time before I set up the batteries etc. to also calibrate gain following your instructions in the README.

The longer explanation is most informative and empowering for me as a beginner in how it all works, and how to do simple things and experiment. Much more than I expected. Having a curious engineering mind I also enjoy learning about how things work. I also appreciate for future readers and myself that you included how to find the most recent release to install.

You ask how I installed OpenHantek6022. I was pleasantly surprised to find it in the Software Manager, so I am guessing it would be a .deb package, but I don't know. I simply clicked install, to get the flatpak version. Then it was 3.2.4, now it 3.2.5, the current one. Very easy (and in hindsight I guess that is why I have done very little with .deb packages, except when following instructions, because more than 95% of things I have needed I have installed with a click in Software Manager, hence being such a novice). As to the permissions I have when installing in this flatpak way, I don't know.

As you write, it really is as easy as 1-2-3 under Linux (as long as you know the steps). I am grateful for what I have learnt through your answer and impressed to get your answer so quickly too, as I am sure you are a busy man. I consider this matter fully answered and closed, but I don't know how this communication platform works, so will leave it with this reply, unless you ask me to do something. 

PS. A little thing I came across was that when trying to run the calibration a second time, to get a config file, I got a lot of error messages. When I shut down the OpenHantek6022 I had opened in the meantime it worked perfectly, so it seems that the API and .py routines don't work when the OpenHantek6022 scope is open. A person who understands what they are doing to higher degree would never make this mistake, but a beginner could and not understand why, thinking something was faulty or had been installed incorrectly. Writing this for the benefit of other beginners  ;o)

**Reply by Ho-Ro** [reply]
Hi @carstensun , thank you for the reply that everything is working - I really need this kind of response to make sure my advices are ok. (Rationnale: I do a lot of development on my computer and therefor I have a lot of libraries and tools installed that are not common on user PCs. So I miss sometimes to mention some needed pre installed libs or tools). For OH6022 I set up an [automatic build action](https://github.com/OpenHantek/OpenHantek6022/actions) here at GitHub that runs every time on some servers somewhere at GitHub when I push some changes (it runs even when I push a typo in the README file). This action builds packages for Linux (using gcc on an Ubuntu system), MacOS (using clang on Mac) and Windows (msvc on W10). So in the end the users get ready-made packages while I see that OH6022 compiles on three different OSs as well as on three different compilers (most of the time the Win compiler makes trouble due to not fully following the C++ standard).

Regarding your issue with parallel access - this is not possible. The important part of the error messages is `usb1.USBErrorBusy: LIBUSB_ERROR_BUSY [-6]`. Only one program can use the scope. I will check if it is possible to install a more intuitive error handling.

Martin

**Reply by Ho-Ro** [reply]
Error handling improved, please check [latest release 2.10.2](https://github.com/Ho-Ro/Hantek6022API/releases/tag/2.10.2).

**Comment by Ho-Ro** [comment]
The latest commit ([unstable](https://github.com/OpenHantek/OpenHantek6022/releases/tag/unstable)) provides a new approach for integrated online calibration regardless of OS:

1. Short circuit the inputs.
2. Enable menu setting "Oscilloscope"/"Calibrate Offset".
3. Select all voltage settings for CH1 and CH2.
4. Disable menu setting "Oscilloscope"/"Calibrate Offset".
The calibration values will be saved on shutdown of the scope.

On next start the calibration will be read and used until next change.
